Rubberduck.VBEEditor/ComManagement/ProjectsRepository.cs

```diff
@@ -0,0 +1,278 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using Rubberduck.VBEditor.Extensions;
+
+namespace Rubberduck.VBEditor.ComManagement
+{
+    public class ProjectsRepository : IProjectsRepository
+    {
+        private IVBProjects _projectsCollection;
+        private readonly IDictionary<string, IVBProject> _projects = new Dictionary<string, IVBProject>();
+        private readonly IDictionary<string, IVBComponents> _componentsCollections = new Dictionary<string, IVBComponents>();
+        private readonly IDictionary<QualifiedModuleName, IVBComponent> _components = new Dictionary<QualifiedModuleName, IVBComponent>();
+        private readonly IDictionary<QualifiedModuleName, ICodeModule> _codeModules = new Dictionary<QualifiedModuleName, ICodeModule>();
+
+        private readonly ReaderWriterLockSlim _refreshProtectionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+
+        public ProjectsRepository(IVBE vbe)
+        {
+            _projectsCollection = vbe.VBProjects;
+            LoadCollections();
+        }
+
+        private void LoadCollections()
+        {
+            LoadProjects();
+            LoadComponentsCollections();
+            LoadComponents();
+        }
+
+        private void LoadProjects()
+        {
+            foreach (var project in _projectsCollection)
+            {
+                EnsureValidProjectId(project);
+                _projects.Add(project.ProjectId, project);
+            }
+        }
+
+        private void EnsureValidProjectId(IVBProject project)
+        {
+            if (string.IsNullOrEmpty(project.ProjectId) || _projects.Keys.Contains(project.ProjectId))
+            {
+                project.AssignProjectId();
+            }
+        }
+
+        private void LoadComponentsCollections()
+        {
+            foreach (var (projectId, project) in _projects)
+            {
+                _componentsCollections.Add(projectId, project.VBComponents);
+            }
+        }
+
+        private void LoadComponents()
+        {
+            foreach (var components in _componentsCollections.Values)
+            {
+                LoadComponents(components);
+            }
+        }
+
+        private void LoadComponents(IVBComponents componentsCollection)
+        {
+            foreach (var component in componentsCollection)
+            {
+                var qmn = new QualifiedModuleName(component);
+                _components.Add(qmn, component);
+                _codeModules.Add(qmn, component.CodeModule);
+            }
+        }
+
+        public void Refresh()
+        {
+            ExecuteWithinWriteLock(() => RefreshCollections());
+        }
+
+        private void ExecuteWithinWriteLock(Action action)
```

Comment:
This is private but I wonder if calling consumers will need to be able to request locks on the collections or members for their action?

Reply:
I would rather not like to hand out a handle to an internal lock to anyone outside the class. Furthermore, the existence of the lock is an implementation detail.